wrapper/src/main/java/software/amazon/jdbc/plugin/customendpoint/CustomEndpointMonitorImpl.java

```diff
@@ -0,0 +1,238 @@
+
+  protected final PluginService pluginService;
+  // TODO: is it problematic for each monitor thread to have its own executor service?
+  protected final ExecutorService monitorExecutor = Executors.newSingleThreadExecutor(runnableTarget -> {
+    final Thread monitoringThread = new Thread(runnableTarget);
+    monitoringThread.setDaemon(true);
+    if (!StringUtils.isNullOrEmpty(monitoringThread.getName())) {
+      monitoringThread.setName(monitoringThread.getName() + "-CustomEndpointMonitor");
+    }
+    return monitoringThread;
+  });
+
+  private final TelemetryCounter infoChangedCounter;
+
+  /**
+   * Constructs a CustomEndpointMonitorImpl instance for the host specified by {@code customEndpointHostSpec}.
+   *
+   * @param pluginService          The plugin service to use to update the plugin service info cache with custom
+   *                               endpoint info.
+   * @param customEndpointHostSpec The host information for the custom endpoint to be monitored.
+   * @param region                 The region of the custom endpoint to be monitored.
+   * @param refreshRateNano        Controls how often the custom endpoint information should be fetched and analyzed for
+   *                               changes. The value specified should be in nanoseconds.
+   * @param rdsClientFunc          The function to call to create the RDS client that will fetch custom endpoint
+   *                               information.
+   */
+  public CustomEndpointMonitorImpl(
+      PluginService pluginService,
+      HostSpec customEndpointHostSpec,
+      Region region,
+      long refreshRateNano,
+      BiFunction<HostSpec, Region, RdsClient> rdsClientFunc) {
+    this.pluginService = pluginService;
+    this.customEndpointHostSpec = customEndpointHostSpec;
+    this.region = region;
+    this.refreshRateNano = refreshRateNano;
+    this.rdsClient = rdsClientFunc.apply(customEndpointHostSpec, this.region);
+
+    this.endpointIdentifier = this.rdsUtils.getRdsClusterId(customEndpointHostSpec.getHost());
+    if (StringUtils.isNullOrEmpty(this.endpointIdentifier)) {
+      throw new RuntimeException(
+          Messages.get(
+              "CustomEndpointMonitorImpl.errorParsingEndpointIdentifier",
+              new Object[] {customEndpointHostSpec.getHost()}));
+    }
+
+    TelemetryFactory telemetryFactory = this.pluginService.getTelemetryFactory();
+    this.infoChangedCounter = telemetryFactory.createCounter(TELEMETRY_ENDPOINT_INFO_CHANGED);
+
+    this.monitorExecutor.submit(this);
+    this.monitorExecutor.shutdown();
+  }
+
+  /**
+   * Analyzes a given custom endpoint for changes to custom endpoint information.
+   */
+  @Override
+  public void run() {
+    LOGGER.fine(
+        Messages.get(
+            "CustomEndpointMonitorImpl.startingMonitor",
+            new Object[] { this.customEndpointHostSpec.getHost() }));
+
+    try {
+      while (!this.stop.get() && !Thread.currentThread().isInterrupted()) {
```

Comment:
Can we rid of `Thread.currentThread().isInterrupted()` condition and replace it with `return`  in corresponding try-catch block?

Reply:
as mentioned above, the return statement isn't necessary because the monitor will already stop if an interrupted exception occurs. I believe an interrupted exception is only thrown if it is blocked (waiting or sleeping) though, which is why I'm checking the condition here.

Reply:
InterruptedException is cough by try-catch block and that leads to stop of the thread. I'm still not sure we need this condition.

Reply:
InterruptedExceptions are only thrown if the thread is interrupted while performing a blocking operation like sleep. Otherwise, if the thread is not blocking, the interrupted flag will be set but no interruption is thrown. If I removed this condition, the thread may continue running normally even after being interrupted.